src/qibo/backends/numpy.py

```diff
@@ -127,6 +127,13 @@ def matrix_fused(self, fgate):
             # small tensor calculations
             # explicit to_numpy see https://github.com/qiboteam/qibo/issues/928
             gmatrix = self.to_numpy(gate.matrix(self))
+            # add controls if controls were instantiated using
+            # the ``Gate.controlled_by`` method
+            num_controls = len(gate.control_qubits)
+            if num_controls > 0:
+                gmatrix = block_diag(
+                    np.eye(2 ** len(gate.qubits) - len(gmatrix)), gmatrix
+                )
```

Comment:
I haven't tested, but are you sure that this is correct in all cases? My main concern is that it is not using the `control_qubits` indices in any way other then length. Shouldn't for example `gmatrix` be different for 
```py
Unitary(matrix, 0).controlled_by(1, 2, 3)
```
compared to 
```py
Unitary(matrix, 2).controlled_by(0, 1, 3)
```
In the second case you would have to put identities on both sides.

Reply:
I have tested this locally before, and it works fine because the only thing the new lines is doing is to get the matrix into the right shape for a gate with controls and target qubits. Then the inclusion of identities and assigment of correct order of qubits is done the same way as before.